cmd/benchttp/main.go

```diff
@@ -0,0 +1,53 @@
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+
+	"github.com/benchttp/runner/config"
+)
+
+var errUsage = errors.New("usage")
+
+func main() {
+	if err := run(); err != nil {
+		fmt.Println(err)
+		if errors.Is(err, errUsage) {
+			flag.Usage()
+		}
+		os.Exit(1)
+	}
+}
+
+func run() error {
+	if len(os.Args) < 2 {
+		return fmt.Errorf("%w: no command specified", errUsage)
+	}
+
+	var cmd command
+	args := os.Args[1:]
+
+	switch sub := args[0]; sub {
+	case "run":
+		cmd = cmdRun{
+			flagset:            flag.NewFlagSet("run", flag.ExitOnError),
+			config:             config.Default(),
+			defaultConfigFiles: defaultConfigFiles,
+		}
```

Comment:
Well technically we could even declare `flagset` inside `cmdRun`, it would work equally.
Here it's just about injecting `cmdRun` dependencies rather than relying on globals so it remains predictable based on its input and more easily testable. It also has the advantage to avoid the need of a `newRunCommand` constructor.

So to me it's far from being the obvious way to go, although I agree `cmdRun` defining its own defaults makes perfect sense, so it's clearly worth considering.

Reply:
Fair enough !

I have a slight preference for a `newRunCommand` returning a ready command, without relying on values injection (ie inlining `config.Default()` and `defaultConfigFiles` inside the constructor).

But both are virtually equal, do what you think is best. The suggestion was worth sharing :)

Reply:
What do you think of 0d50cbdfc9849508d93f49da66578fb08fb60975`Set cmdRun defaults in method init`?

It's a bit of a compromise:
- defaults are set internally, so we lose a bit of visibility about our defaults but it clears the package
- no `new` constructor but an `init` method called in `execute`: we lose in terms of testability but no new declaration in the package

Fine for me. Better or worse for you?

Reply:
Seems good to me too ! If we go this way, I would go down the route and make a constructor honestly.
It also removes the "ensure implements interface stuff".

```go
// command is the interface that all benchttp subcommands must implement.
type command interface {
	execute(args []string) error
}

// ensure cmdRun implements command
// var _ command = (*cmdRun)(nil) <- useless

func newRunCommand(fs *flag.FlagSet) command {
	return &cmdRun{
		flagset: fs,
		config:  config.Default(),
		defaultConfigFiles: []string{"./.benchttp.yml", "./.benchttp.yaml", "./.benchttp.json"},
	}
}

func run () {
	var cmd command
	args := os.Args[1:]

	switch sub := args[0]; sub {
	case "run":
		cmd = newRunCommand(flag.NewFlagSet("run", flag.ExitOnError))
	// ...
	}
}
```

> defaults are set internally, so we lose a bit of visibility about our defaults but it clears the package

Well technically true, but I don't think that reading the next struct literal gives us much more info than a constructor: 

```go
cmd = &cmdRun{
	flagset:            flag.NewFlagSet("run", flag.ExitOnError),
	config:             config.Default(),
	defaultConfigFiles: defaultConfigFiles,
}
```

In any case, I do not require any changes or press anything, really go for what you feel is the best.

Reply:
>  If we go this way, I would go down the route and make a constructor honestly.

To me the constructor is not the logical next step here. I don't get the necessity of it, and it'll add either many unnecessary declarations with commands that need no setup (`help`, `version`...?) or inconsistency in their usage  if we decide not to use constructors for them.
`var _ itf = (*impl)(nil)` has always been anecdotic, we use it for immediate feedback near the implementation, but even without a bad implementation would break at line `cmd = cmdRun{}` 🤷‍♂️ 

> Well technically true, but I don't think that reading the next struct literal gives us much more info than a constructor

Yes, I was referring to the package-level `defaultConfigFiles` at the top of the file 😅